### PR TITLE
Populate last terminated container information

### LINKED ISSUE #4919: Kubelet should report last terminated container status
Currently kubelet reports current status only. If pod's restart policy is not never, the last terminated container status will be override by the latest status, and lost. We should report last terminated container status. 

I am going to update our v1beta3 api first, but will hold the implementation until we have #4561

### PR DESCRIPTION
**Pull request by dchen1107**
With this PR, on node:

```
# docker ps -a | grep etcd-server-kubernete
be7e499dbc45        gcr.io/google_containers/etcd:2.0.9                  "/usr/local/bin/etcd   4 hours ago         Up 4 hours                                      k8s_etcd-container.1e717455_etcd-server-kubernetes-master_default_1e5f2c4ea3b621474a42261fac0c1a3d_950d2f78            
91d3df08509e        gcr.io/google_containers/etcd:2.0.9                  "/usr/local/bin/etcd   21 hours ago        Exited (137) 4 hours ago                        k8s_etcd-container.1e717455_etcd-server-kubernetes-master_default_1e5f2c4ea3b621474a42261fac0c1a3d_3d965d46            
7c6274cac6d9        gcr.io/google_containers/etcd:2.0.9                  "/usr/local/bin/etcd   22 hours ago        Exited (137) 21 hours ago                       k8s_etcd-container.1e717455_etcd-server-kubernetes-master_default_1e5f2c4ea3b621474a42261fac0c1a3d_51d1f8ce            
2ca37132942f        gcr.io/google_containers/etcd:2.0.9                  "/usr/local/bin/etcd   28 hours ago        Exited (137) 22 hours ago                       k8s_etcd-container.1e717455_etcd-server-kubernetes-master_default_1e5f2c4ea3b621474a42261fac0c1a3d_3a7ff2c0            
6da0ee59f2b1        gcr.io/google_containers/pause:0.8.0                 "/pause"               28 hours ago        Up 28 hours                                     k8s_POD.1a1fe303_etcd-server-kubernetes-master_default_1e5f2c4ea3b621474a42261fac0c1a3d_e268f17e                       
```

From client:

```
$ cluster/kubectl.sh get -o json pods etcd-server-kubernetes-master
...
    "status": {
        "phase": "Running",
        "Condition": [
            {
                "type": "Ready",
                "status": "True"
            }
        ],
        "containerStatuses": [
            {
                "name": "etcd-container",
                "state": {
                    "running": {
                        "startedAt": "2015-04-09T17:12:43Z"
                    }
                },
                "lastState": {
                    "termination": {
                        "exitCode": 137,
                        "startedAt": "2015-04-09T00:14:15Z",
                        "finishedAt": "2015-04-09T17:12:37Z",
                        "containerID": "docker://91d3df08509e8d31460c89806b5ab84938cb3a6dc0f7d93d435d549cc6fd9c14"
                    }
                },
                "ready": true,
                "restartCount": 3,
                "image": "gcr.io/google_containers/etcd:2.0.9",
                "imageID": "docker://b6b9a86dc06aa1361357ca1b105feba961f6a4145adca6c54e142c0be0fe87b0",
                "containerID": "docker://be7e499dbc456c52b4542a6017852a6bd32bab006160cc1f892e5a6896a18e1d"
            }
        ]
    }

```

Fix #4919

### CODE REVIEW

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -215,6 +215,14 @@ func (self *DockerManager) GetPodStatus(pod *api.Pod) (*api.PodStatus, error) {
 		}
 		// We assume docker return us a list of containers in time order
 		if containerStatus, found := statuses[dockerContainerName]; found {
+			// Populate last termination state
+			if containerStatus.LastTerminationState.Termination == nil {
+				result := self.inspectContainer(value.ID, dockerContainerName, terminationMessagePath)
+				if result.err == nil && result.status.State.Termination != nil {
+					containerStatus.LastTerminationState = result.status.State
+					containerStatus.LastTerminationState.Termination.ContainerID = result.status.ContainerID
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/manager.go]: We should populate this in inspectContainer so we only do it in one place.

File: pkg/kubelet/kubelet_test.go [hunk 2]
```diff
@@ -3657,3 +3657,145 @@ func TestSyncPodsWithRestartPolicy(t *testing.T) {
...
+				}
+			}
+			sort.StringSlice(terminatedContainers).Sort()
+			sort.StringSlice(tt.lastTerminations).Sort()
+			if !reflect.DeepEqual(terminatedContainers, tt.lastTerminations) {
+				t.Errorf("Expected(sorted): %#v, Actual(sorted): %#v", tt.lastTerminations, terminatedContainers)
+			}
+		}
+
+		// 'stop' is because the pod infra container is killed when no container is running.
```
Comment (vmarmol) [inline pkg/kubelet/kubelet_test.go]: WDYT of not checking this and only checking pod status? Should keep the test smaller and more focused. Do we event need to verifyCalls?

**Comment by vmarmol** [comment]
LGTM overall, just some comments.

File: pkg/kubelet/kubelet_test.go (hunk 2 above)
Comment (yujuhong) [inline pkg/kubelet/kubelet_test.go]: +1 for reducing the overly used verifyCalls.

File: pkg/kubelet/dockertools/manager.go (hunk 1 above)
Comment (dchen1107) [inline pkg/kubelet/dockertools/manager.go]: Done.

File: pkg/kubelet/kubelet_test.go (hunk 2 above)
Comment (dchen1107) [inline pkg/kubelet/kubelet_test.go]: Feel the same pain. Removed those non-sense.

**Comment by vmarmol** [comment]
LGTM, thanks @dchen1107!

**Comment by yujuhong** [comment]
@dchen1107, a general question about the container statuses, and whether it should be persisted.
We do container garbage collection, so information such as LastTerminateState may eventually be lost. Even though it has been written in the pod status, because kubelet periodically computes the status from the existing containers (dead or alive), the pod status would later be overwritten with empty information. Is this expected/allowed (that we clear the meaningful information in the pod status because we can no longer reproduce it)?

**Comment by vmarmol** [comment]
Nice catch @yujuhong! We could alleviate this by requiring a minimum of two last statuses for running nodes. This may be desirable given that we'll need that status to calculate things like restart policies anyways. I'm guessing eventually we'll want to checkpoint that information instead of keeping it around in a potentially heavy container.

**Comment by dchen1107** [comment]
That is expected. For v1.0, we agreed at best effort to provide last terminated container information and restart count. Without more comprehensive disk management, we cannot guarantee last terminated container will be reserved on node when node reaches the thresholds. But even with disk management, we still cannot guarantee unless kubelet won't be overcommitted. As our internal system, we might lost last terminated task information.

**Comment by vmarmol** [comment]
SGTM, merging.

**Pull request merged**